### PR TITLE
[WIP] Transition to Typescript

### PR DESCRIPTION
**Pull request by DaSchTour**
Just to track progress here. This will be a lot of work, but just from starting I see a lot of possible problems in the current JS codebase.

1. There are internal functions not described correctly.
2. The Plugin class isn't used correctly by all plugins

But that said it's worth the effort to move to TS to find all this hidden possible problems.

### CODE REVIEW

**Comment by IamManchanda** [comment]
Nice... Happy to contribute 
But cc @kball fot this feature request 

FWIW I support Daniel on this PR

**Comment by IamManchanda** [comment]
FWIW, I will repeat my comment here

> Thanks to the type system, TypeScript helps a lot in minimizing bugs and unit tests.
> and, yeah this is not just an angular thing
> 
> - https://angular.io/guide/quickstart
> - https://www.typescriptlang.org/docs/handbook/react-&-webpack.html
> - https://vuejs.org/v2/guide/typescript.html
> 
> { PS: I haven't tried angular yet, and I am not learning TypeScript for learning angular but for being better in JavaScript }
> 

---

Also, Vue.js is also giving [preference](https://medium.com/the-vue-point/upcoming-typescript-changes-in-vue-2-5-e9bd7e2ecf08) to TypeScript now.
Also, in their roadmap

> **On the Roadmap: TypeScript Support in vue-cli**
After 2.5, we are planning to introduce official support for TypeScript in the next version of vue-cli in order to make it easier for TS+Vue users to kick off new projects. Stay tuned!

**Comment by kball** [comment]
Woah! There's a ton of work in here @DaSchTour, nice work!

I think this is a very interesting direction, and as you point out opens up all sorts of better static analysis and will help us find more things. 

So long as we are able to maintain a straightforward migration path for the various ways folks are using older versions of the framework, I'm game.

**Comment by IamManchanda** [comment]
@kball @DaSchTour ... what's the roadmap on this?
6.5? 7.0?

**Comment by DaSchTour** [comment]
@IamManchanda Depends on how fast I'll get the work done and if there will be a complete rewrite for 7.0

**Comment by IamManchanda** [comment]
Moving to a different language shouldn't be done in the same version series.
Yes typescript is better undoubtedly and should be used but not before **v7**

FWIW... I am adding a label @ v7

**Comment by IamManchanda** [comment]
Also, for v7 rewrite, you should be considering this PR too #10544

File: js/entries/foundation-plugins.d.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+export { Abide } from '../foundation.abide';
```
Comment (IamManchanda) [inline js/entries/foundation-plugins.d.ts]: > js/entries/foundation-plugins.d.ts

Shouldn't the new folder be ts

Comment (DaSchTour) [reply js/entries/foundation-plugins.d.ts]: for compatibility I think we should not rename everything

**Comment by Owlbertz** [comment]
@DaSchTour I would consider myself rather experienced with TS. If you want to delegate something, feel free to get in touch with me.

**Comment by DaSchTour** [comment]
@Owlbertz I'll possibly won't have time to do anything on this topic till and and of this year. So you could try to go on with this as you like. Fix the TS errors that are currently poping up. Add the typescript builder to webpack/gulp. And I guess backporting of JS changes from other PRs also needs to be done.

**Comment by IamManchanda** [comment]
One fun (little unrelated) stuff,
JavaScript wrapped with **Typescript** not only let you do the web
but also make you at good place to learn ahead cross platform **modern** mobile development!

Here are quick links I found out:

- [Swift is like TypeScript](https://alhazmy13.github.io/swift-is-like-typescript/
- )
- [Kotlin is like TypeScript](https://gi-no.github.io/kotlin-is-like-typescript/)
- [Swift is like Kotlin](http://nilhcem.com/swift-is-like-kotlin/)

So it's a good language that should be learn today if you are doing JavaScript

File: js/entries/foundation-plugins.d.ts (hunk 1 above)
Comment (IamManchanda) [reply js/entries/foundation-plugins.d.ts]: Yeah same like how kotlin does with java
Gotcha ... Agree!

**Comment by IamManchanda** [comment]
V7 requests are being closed for the time being. We have them labeled so we can readdress them in the future.

**Comment by DaSchTour** [comment]
Well. I don't know why this was labeled V7? But that was never meant to be only for V7.
But as I don't have time for this maybe it's better to not do TypeScript at Foundation. We already stoped using Foundation JS for new projects.

**Pull request closed without merging**